src/LodeRunner/src/Subscribers/ClientStatusUpdater.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using LodeRunner.Core.Events;
+using LodeRunner.Core.Models;
+using LodeRunner.Data.Interfaces;
+
+namespace LodeRunner.Subscribers
+{
+    internal class ClientStatusUpdater : IDisposable
+    {
+        private const int MaxRetryAttempts = 3;
+        private readonly ClientStatus clientStatus;
+        private readonly IClientStatusService clientStatusService;
+        private int failures = 0;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ClientStatusUpdater"/> class.
+        /// </summary>
+        /// <param name="clientStatusService">The client status service.</param>
+        /// <param name="clientStatus">The client status.</param>
+        /// <param name="cancellationTokenSource">The cancellation token source.</param>
+        public ClientStatusUpdater(IClientStatusService clientStatusService, ClientStatus clientStatus)
+        {
+            this.clientStatus = clientStatus;
+            this.clientStatusService = clientStatusService;
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose()
+        {
+            // Dispose any objects here.
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Updates the cosmos status.
+        /// </summary>
+        /// <param name="sender">The sender.</param>
+        /// <param name="args">The <see cref="ClientStatusEventArgs"/> instance containing the event data.</param>
+        public async void UpdateCosmosStatus(object sender, ClientStatusEventArgs args)
+        {
+            // TODO: do we need a lock here?
+            this.clientStatus.Message = args.Message;
+            this.clientStatus.Status = args.Status;
+
+            try
+            {
+                _ = await this.clientStatusService.Post(this.clientStatus, args.CancelTokenSource.Token).ConfigureAwait(false);
+            }
```

Comment:
On a successful update, this should reset that error count to 0.  The idea being that we won't care about transient failures on the status update, but if we have N (currently 3) consecutive then we take it as a meaningful pattern and cancel the operation.

Probably just add a direct assignment:
```csharp
            try
            {
                _ = await this.clientStatusService.Post(this.clientStatus, args.CancelTokenSource.Token).ConfigureAwait(false);
               // reset failure count on success
                failures = 0;
            }
```

Reply:
I totally missed that part of the requirements, thanks four pointing that out.